src/main/java/com/amazon/opendistroforelasticsearch/ad/rest/handler/ValidateAnomalyDetectorActionHandler.java

```diff
@@ -0,0 +1,590 @@
+        client
+            .search(
+                searchRequest,
+                ActionListener.wrap(searchResponse -> onQueryFilterSearch(searchResponse), exception -> { onFailure(exception); })
+            );
+    }
+
+    private void onQueryFilterSearch(SearchResponse response) throws IOException {
+        if (response.getHits().getTotalHits().value <= 0) {
+            List<String> filterError = new ArrayList<>();
+            filterError
+                .add(
+                    "query filter is potentially wrong as no hits were found at all or no historical data in last "
+                        + NUM_OF_INTERVALS_CHECKED_FILTER
+                        + " intervals"
+                );
+            suggestedChangesMap.put("filter_query", filterError);
+            validateAnomalyDetectorResponse();
+        } else {
+            long timestamp = (long) response.getHits().getHits()[0].getSourceAsMap().get("timestamp");
+            featureQueryValidation(timestamp);
+        }
+    }
+
+    private void featureQueryValidation(long startTimeStamp) throws IOException {
+        long delayMillis = Optional
+            .ofNullable((IntervalTimeConfiguration) anomalyDetector.getWindowDelay())
+            .map(t -> t.toDuration().toMillis())
+            .orElse(0L);
+        long[] startEnd = startEndTimeRangeWithIntervals(NUM_OF_INTERVALS_CHECKED);
+        long dataEndTime = startEnd[1];
+        long dataStartTime = startEnd[0];
+        if (startEnd[0] > startTimeStamp) {
+            dataStartTime = startTimeStamp;
+        }
+        IntervalTimeConfiguration searchRange = new IntervalTimeConfiguration(FEATURE_VALIDATION_TIME_BACK_MINUTES, ChronoUnit.MINUTES);
+        long searchRangeTime = Optional.ofNullable(searchRange).map(t -> t.toDuration().toMillis()).orElse(0L);
+        long startTimeWithSetTime = startEnd[1] - (searchRangeTime - delayMillis);
+        // Make sure start time includes timestamp seen by filter query check
+        if (startEnd[0] > startTimeWithSetTime) {
+            dataStartTime = startTimeWithSetTime;
+        }
+        AtomicInteger featureCounter = new AtomicInteger();
+        RangeQueryBuilder rangeQuery = new RangeQueryBuilder(anomalyDetector.getTimeField())
+            .from(dataStartTime)
+            .to(dataEndTime)
+            .format("epoch_millis")
+            .includeLower(true)
+            .includeUpper(false);
+        if (anomalyDetector.getFeatureAttributes() != null) {
+            for (Feature feature : anomalyDetector.getFeatureAttributes()) {
+                ParseUtils.parseAggregators(feature.getAggregation().toString(), xContent, feature.getId());
+                XContentParser parser = XContentType.JSON
+                    .xContent()
+                    .createParser(xContent, LoggingDeprecationHandler.INSTANCE, feature.getAggregation().toString());
+                parser.nextToken();
+                List<String> fieldNames = parseAggregationRequest(parser);
```

Comment:
You can encapsulate such logic into a separate method `getFieldNamesForFeature(Feature feature)`, so that `getFeatureFieldNames()` can just use `getFieldNamesForFeature` to get fieldNames for each feature

Reply:
done